File: examples/sensor_fusion.py
"""
sensor_fusion.py — 二值逻辑传感器融合（对比实现）
功能与 sensor_fusion.san 相同，但用 Python 的 None/枚举处理不确定性
展示二值逻辑需要多少额外代码来处理三值逻辑天然支持的场景
"""
import random
from enum import Enum
from typing import Optional, Tuple

class SensorState(Enum):
    NORMAL = "normal"
    FAULT = "fault"
    OFFLINE = "offline"

def simulate_sensor(name: str, fault_rate: int) -> SensorState:
    r = random.randint(1, 100)
    if r <= fault_rate:
        return SensorState.FAULT
    if r <= fault_rate + 10:
        return SensorState.OFFLINE
    return SensorState.NORMAL

def read_temperature() -> Tuple[SensorState, Optional[int]]:
    state = simulate_sensor("温度", 15)
    if state == SensorState.NORMAL:
        return (state, random.randint(18, 35))
    elif state == SensorState.OFFLINE:
        return (state, None)  # 离线返回 None
    else:
        return (state, None)  # 故障返回 None

def read_humidity() -> Tuple[SensorState, Optional[int]]:
    state = simulate_sensor("湿度", 10)
    if state == SensorState.NORMAL:
        return (state, random.randint(30, 80))
    return (state, None)

def read_gas() -> Tuple[SensorState, Optional[int]]:
    state = simulate_sensor("气体", 20)
    if state == SensorState.NORMAL:
        return (state, random.randint(0, 500))
    return (state, None)

def fuse_sensor_states(states: list[SensorState]) -> SensorState:
    """融合多个传感器状态 — 需要显式枚举处理"""
    has_fault = any(s == SensorState.FAULT for s in states)
    has_offline = any(s == SensorState.OFFLINE for s in states)
    all_normal = all(s == SensorState.NORMAL for s in states)

    if has_fault:
        return SensorState.FAULT
    if all_normal:
        return SensorState.NORMAL
    return SensorState.OFFLINE  # 至少一个离线

def temperature_safe(temp: Optional[int]) -> SensorState:
    if temp is None:
        return SensorState.OFFLINE
    if temp > 40 or temp < 10:
        return SensorState.FAULT
    return SensorState.NORMAL

def humidity_safe(humid: Optional[int]) -> SensorState:
    if humid is None:
        return SensorState.OFFLINE
    if humid > 90 or humid < 20:
        return SensorState.FAULT
    return SensorState.NORMAL

def gas_safe(gas: Optional[int]) -> SensorState:
    if gas is None:
        return SensorState.OFFLINE
    if gas > 400:
        return SensorState.FAULT
    if gas > 200:
        return SensorState.OFFLINE  # 警告
    return SensorState.NORMAL

def environment_decision(overall: SensorState, temp, humid, gas) -> str:
    if overall == SensorState.FAULT:
        print(f"🚨 警报：传感器故障或环境危险！")
        print(f"  温度: {temp}  湿度: {humid}  气体: {gas}")
        return "紧急停机"
    elif overall == SensorState.OFFLINE:
        print("⚠️ 警告：部分传感器离线，降级运行")
        if temp is not None and temp > 35:
            return "降级-开启风扇"
        return "降级-维持现状"
    else:
        print("✅ 环境正常，全速运行")
        if temp is not None and temp > 30:
            return "开启空调"
        if humid is not None and humid < 40:
            return "开启加湿器"
        return "维持现状"

def main():
    print("═" * 43)
    print("  二值逻辑传感器融合系统 v1.0 (Python)")
    print("═" * 43)
    print()

    count_normal = 0
    count_uncertain = 0
    count_abnormal = 0

    for round_num in range(1, 11):
        print(f"--- 第 {round_num} 轮检测 ---")

        temp_state, temp_val = read_temperature()
        humid_state, humid_val = read_humidity()
        gas_state, gas_val = read_gas()

        temp_safe = temperature_safe(temp_val)
        humid_safe = humidity_safe(humid_val)
        gas_safe_val = gas_safe(gas_val)

        sensor_fusion = fuse_sensor_states([temp_state, humid_state, gas_state])
        safety_fusion = fuse_sensor_states([temp_safe, humid_safe, gas_safe_val])

        # 最终状态：需要显式处理两个枚举的组合
        if sensor_fusion == SensorState.FAULT or safety_fusion == SensorState.FAULT:
            final = SensorState.FAULT
        elif sensor_fusion == SensorState.NORMAL and safety_fusion == SensorState.NORMAL:
            final = SensorState.NORMAL
        else:
            final = SensorState.OFFLINE

        print(f"  传感器: 温度={temp_state.value} 湿度={humid_state.value} "
              f"气体={gas_state.value} → 融合={sensor_fusion.value}")
        print(f"  安全性: 温度={temp_safe.value} 湿度={humid_safe.value} "
              f"气体={gas_safe_val.value} → 融合={safety_fusion.value}")

        decision = environment_decision(final, temp_val, humid_val, gas_val)
        print(f"  决策: {decision}")
        print()

        if final == SensorState.NORMAL:
            count_normal += 1
        elif final == SensorState.OFFLINE:
            count_uncertain += 1
        else:
            count_abnormal += 1

    print("═" * 43)
    print("  统计汇总")
    print("═" * 43)
    print(f"  正常: {count_normal} 轮")
    print(f"  不确定: {count_uncertain} 轮")
    print(f"  异常: {count_abnormal} 轮")
    print()
    print("二值逻辑需要：Enum 类定义、None 检查、显式状态组合")
    print("三值逻辑只需：真/假/可能 + 且/或/非 运算符")

if __name__ == "__main__":
    main()

File: examples/sensor_fusion.c
/*
 * sensor_fusion.c — 二值逻辑传感器融合（对比实现）
 * 功能与 sensor_fusion.san 相同
 * 展示 C 语言处理三态需要多少额外代码
 */
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <stdbool.h>

/* C 语言没有三值类型，必须用枚举 + switch */
typedef enum { STATE_TRUE, STATE_FALSE, STATE_MAYBE } tristate_t;

typedef struct {
    tristate_t state;
    int value;
} sensor_reading_t;

/* 三值 AND — C 需要手写，三言内置 */
tristate_t tri_and(tristate_t a, tristate_t b) {
    if (a == STATE_FALSE || b == STATE_FALSE) return STATE_FALSE;
    if (a == STATE_TRUE && b == STATE_TRUE) return STATE_TRUE;
    return STATE_MAYBE;
}

/* 三值 OR — C 需要手写 */
tristate_t tri_or(tristate_t a, tristate_t b) {
    if (a == STATE_TRUE || b == STATE_TRUE) return STATE_TRUE;
    if (a == STATE_FALSE && b == STATE_FALSE) return STATE_FALSE;
    return STATE_MAYBE;
}

tristate_t simulate_sensor(const char *name, int fault_rate) {
    int r = rand() % 100 + 1;
    if (r <= fault_rate) return STATE_FALSE;
    if (r <= fault_rate + 10) return STATE_MAYBE;
    return STATE_TRUE;
}

sensor_reading_t read_temperature(void) {
    tristate_t state = simulate_sensor("temp", 15);
    sensor_reading_t r = {state, 0};
    if (state == STATE_TRUE) r.value = rand() % 18 + 18;
    else if (state == STATE_MAYBE) r.value = 0;
    else r.value = -999;
    return r;
}

sensor_reading_t read_humidity(void) {
    tristate_t state = simulate_sensor("humid", 10);
    sensor_reading_t r = {state, 0};
    if (state == STATE_TRUE) r.value = rand() % 51 + 30;
    else r.value = -999;
    return r;
}

sensor_reading_t read_gas(void) {
    tristate_t state = simulate_sensor("gas", 20);
    sensor_reading_t r = {state, 0};
    if (state == STATE_TRUE) r.value = rand() % 501;
    else r.value = -999;
    return r;
}

tristate_t fuse_states(tristate_t *states, int n) {
    /* C 需要循环 + 显式枚举比较 */
    bool has_fault = false, has_maybe = false, all_true = true;
    for (int i = 0; i < n; i++) {
        if (states[i] == STATE_FALSE) has_fault = true;
        if (states[i] == STATE_MAYBE) has_maybe = true;
        if (states[i] != STATE_TRUE) all_true = false;
    }
    if (has_fault) return STATE_FALSE;
    if (all_true) return STATE_TRUE;
    return STATE_MAYBE;
}

tristate_t temperature_safe(int temp) {
    if (temp == -999) return STATE_MAYBE;
    if (temp > 40 || temp < 10) return STATE_FALSE;
    return STATE_TRUE;
}

tristate_t humidity_safe(int humid) {
    if (humid == -999) return STATE_MAYBE;
    if (humid > 90 || humid < 20) return STATE_FALSE;
    return STATE_TRUE;
}

tristate_t gas_safe(int gas) {
    if (gas == -999) return STATE_MAYBE;
    if (gas > 400) return STATE_FALSE;
    if (gas > 200) return STATE_MAYBE;
    return STATE_TRUE;
}

const char* state_str(tristate_t s) {
    switch (s) {
        case STATE_TRUE: return "真";
        case STATE_FALSE: return "假";
        case STATE_MAYBE: return "可能";
    }
    return "?";
}

void environment_decision(tristate_t overall, int temp, int humid, int gas) {
    switch (overall) {
        case STATE_FALSE:
            printf("🚨 警报：传感器故障或环境危险！\n");
            printf("  温度: %d  湿度: %d  气体: %d\n", temp, humid, gas);
            printf("  决策: 紧急停机\n");
            break;
        case STATE_MAYBE:
            printf("⚠️ 警告：部分传感器离线，降级运行\n");
            if (temp > 35) printf("  决策: 降级-开启风扇\n");
            else printf("  决策: 降级-维持现状\n");
            break;
        case STATE_TRUE:
            printf("✅ 环境正常，全速运行\n");
            if (temp > 30) printf("  决策: 开启空调\n");
            else if (humid < 40) printf("  决策: 开启加湿器\n");
            else printf("  决策: 维持现状\n");
            break;
    }
}

int main(void) {
    srand((unsigned)time(NULL));

    printf("═══════════════════════════════════════\n");
    printf("  二值逻辑传感器融合系统 v1.0 (C)\n");
    printf("═══════════════════════════════════════\n\n");

    int count_normal = 0, count_uncertain = 0, count_abnormal = 0;

    for (int round = 1; round <= 10; round++) {
        printf("--- 第 %d 轮检测 ---\n", round);

        sensor_reading_t temp = read_temperature();
        sensor_reading_t humid = read_humidity();
        sensor_reading_t gas = read_gas();

        tristate_t temp_safe = temperature_safe(temp.value);
        tristate_t humid_safe = humidity_safe(humid.value);
        tristate_t gas_s = gas_safe(gas.value);

        tristate_t sensor_states[] = {temp.state, humid.state, gas.state};
        tristate_t safety_states[] = {temp_safe, humid_safe, gas_s};

        tristate_t sensor_fusion = fuse_states(sensor_states, 3);
        tristate_t safety_fusion = fuse_states(safety_states, 3);
        tristate_t final_state = tri_and(sensor_fusion, safety_fusion);

        printf("  传感器: 温度=%s 湿度=%s 气体=%s → 融合=%s\n",
               state_str(temp.state), state_str(humid.state),
               state_str(gas.state), state_str(sensor_fusion));
        printf("  安全性: 温度=%s 湿度=%s 气体=%s → 融合=%s\n",
               state_str(temp_safe), state_str(humid_safe),
               state_str(gas_s), state_str(safety_fusion));

        environment_decision(final_state, temp.value, humid.value, gas.value);
        printf("\n");

        switch (final_state) {
            case STATE_TRUE: count_normal++; break;
            case STATE_MAYBE: count_uncertain++; break;
            case STATE_FALSE: count_abnormal++; break;
        }
    }

    printf("═══════════════════════════════════════\n");
    printf("  统计汇总\n");
    printf("═══════════════════════════════════════\n");
    printf("  正常: %d 轮\n", count_normal);
    printf("  不确定: %d 轮\n", count_uncertain);
    printf("  异常: %d 轮\n", count_abnormal);
    printf("\n");
    printf("C 语言需要：enum 定义、手写 tri_and/tri_or、switch-case 处理\n");
    printf("三言只需：真/假/可能 + 且/或/非 运算符\n");
    return 0;
}
